[Drape] Introduce developer sandbox app for rendering debugging and developement

renderexpert opened the pull request:
# Current problem
The existing Qt-based desktop app supports only OpenGL rendering, while mobile clients use Metal and Vulkan mostly. More than that, the Qt-based app injects its own OpenGL code and mixes rendering with Qt's OpenGL. It increases the distance between desktop and mobile clients even more. Thus, the development and debugging of graphics features can be reliably done only on mobile devices. It's much more tricky because of the lack of good tooling, especially for Vulkan.
The latest versions of Qt provide support for Metal and Vulkan on desktop (not supported by OMaps at the moment, though). However, Qt still does its own Vulkan/Metal API calls there and requires a lot of additional code to be compatible with OMaps approach - execute business logic in the main thread and do rendering in a separate thread(s). 

# My approach
There are two popular and well-known libraries in the community of game and rendering developers
1. GLFW - https://github.com/glfw/glfw
2. Dear imGui - https://github.com/ocornut/imgui

These libraries do OS-level window management and simplistic GUI for tons of graphics demos and games. I created a separate app in the OMaps repository that utilises them.

How it looks (click to open video):
[![video 1](https://img.youtube.com/vi/2o73QoAqZFA/0.jpg)](https://www.youtube.com/watch?v=2o73QoAqZFA)

[![video 2](https://img.youtube.com/vi/cy88pk9tPW0/0.jpg)](https://www.youtube.com/watch?v=cy88pk9tPW0)

# Features
1. All rendering is done by the Drape, no external graphics API calls
2. Simplistic UI allows to do map manipulations, to download maps, to work with location, to enable hidden rendering features (like SMAA and POI displacement debugging visualisation) and can be easily extended for other features in the future (3D, routing, elevation maps and so on)
3. Graphics API can be switched without app re-run! (It is hard to implement the same in Qt)
4. All API works on macOS (Vulkan via MoltenVK - https://github.com/KhronosGroup/MoltenVK)
5. Vulkan and OpenGL work natively on Linux (tested on Ubuntu)

# How to use
Just build and run `dev_sandbox` project

#

AndrewShkrob commented:
What's the point of this tool?
I mean, I understand the point but I'm not sure if it's worth it. We already have a desktop app and I believe it would be better to implement this functionality (different graphics API) in it. Even if it's harder. And I don't think we need to switch apis in runtime. It's of course good to have it but if it requires another tool then maybe not :)
Now we'll have one more tool to support.

Even if we don't "officially" support desktop it's still good to have. And we have people who use it. So, we won't abandon qt_desktop anyway.

What am I missing?

renderexpert commented:
@AndrewShkrob  Well, as it's written it's for rendering debugging and developement. If you don't do it, possibly it may be useless for you. If you need to check how the same place renders in Vulkan and Metal, you value switching APIs in runtime.
Overall, I undestand sentitment "why do we need one more tool", my position here that 
1. This tool can do what current qt app can't and who knows when and if it will be implemented
2. This tool uses the same approach for rendering as mobile apps what Qt app will never do
3. This tool does not replace Qt app, so no one who got used to Qt app suffers
4. Maitanance issue is exaggerated IMO. This tool 90% consists of the Drape engine, so the maintanace here is mainatance of the Drape engine amount of what is not really increased by this tool. More than that I dare to say it decreases by it. I found and fixed at least 2 bugs in Vulkan while I implemented it

renderexpert commented:
@vng Could you please review it

vng commented on dev_sandbox/main.cpp:
```diff
@@ -0,0 +1,672 @@
+std::function<void(double x, double y)> onMouseMove;
+std::function<void(double x, double y, double xOffset, double yOffset)> onScroll;
+std::function<void(int key, int scancode, int action, int mods)> onKeyboardButton;
+std::function<void(float xscale, float yscale)> onContentScale;
```
nit: I'd prefer to make one Window or App instance here that holds all these callbacks. To have one global object instead of many.

vng commented on dev_sandbox/main.cpp:
```diff
@@ -0,0 +1,672 @@
+#if defined(OMIM_OS_MAC)
+    x *= visualScale;
+    y *= visualScale;
+#endif
```
This stuff here and above is a bit mind-blowing :) Why do we multiply coordinates on Mac and divide on Linux.

renderexpert replied:
Unfortunately, OS are not consistent in terms of visual scale in glfw library (possibly it's native behaviour on this OS). macOS returns width/height in the same units regardless of screen scaling, while on Linux it returns width/height premultiplied to display scaling.

vng commented on dev_sandbox/main_linux.cpp:
```diff
@@ -0,0 +1,411 @@
+                 bool usePixelBuffer)
+    : m_glx(glx), m_display(display), m_window(window)
+  {
+    static int visualAttribs[] = {
```
nit: I suppose static is redundant here

AndrewShkrob commented on dev_sandbox/CMakeLists.txt:
```diff
@@ -0,0 +1,144 @@
+  copy_resources(icudt75l.dat)
+endif()
+
+if (PLATFORM_MAC)
+  execute_process(
+    COMMAND cp -r ${OMIM_ROOT}/tools/shaders_compiler/macos ${RESOURCES_FOLDER}/shaders_compiler
+  )
+elseif (PLATFORM_LINUX)
+  execute_process(
+    COMMAND cp -r ${OMIM_ROOT}/tools/shaders_compiler/linux ${RESOURCES_FOLDER}/shaders_compiler
+  )
+endif()
```
Is it needed? Shaders compiler is used only in tests

AndrewShkrob commented on dev_sandbox/CMakeLists.txt:
```diff
@@ -0,0 +1,144 @@
+  endforeach()
+endfunction()
+
+copy_resources(
```
How about vulkan shaders folder? Is the app going to support Vulkan?

renderexpert replied:
Good catch!

rtsisyk reviewed (approved):
LGTM

pastk commented:
It'd be great to add some docs - what is it, how to build and run. Also please add to `docs/STRUCTURE.md`.

pastk commented:
I have build issues on Ubuntu 24.04.

Fresh build:
```
tools/unix/build_omim.sh -c -r dev_sandbox
...
[113/113] Linking CXX executable OMapsDevSandbox
FAILED: OMapsDevSandbox 
: && /usr/bin/ccache /usr/bin/c++ -O3 -DNDEBUG -fuse-ld=gold    -fdiagnostics-color=always dev_sandbox/CMakeFiles/dev_sandbox.dir/Unity/unity_0_cxx.cxx.o -o OMapsDevSandbox  libgenerator.a  -lqt_common  libmap.a  libgflags_nothreads.a  libglfw3.a  libimgui.a  /usr/lib/x86_64-linux-gnu/libvulkan.so  libtess2.a  librouting.a  libdescriptions.a  libkml.a  libdrape_frontend.a  libtransit.a  libtraffic.a  librouting_common.a  libshaders.a  libdrape.a  libsearch.a  libindexer.a  libeditor.a  libsearch.a  libindexer.a  libeditor.a  libstorage.a  libmwm_diff.a  libbsdiff.a  libge0.a  libopenlocationcode.a  libprotobuf.a  libopening_hours.a  libpugixml.a  libplatform.a  libcoding.a  libcppjansson.a  3party/jansson/jansson/lib/libjansson.a  libexpat.a  -lm  libsuccinct.a  /usr/lib/x86_64-linux-gnu/libicuuc.so  libminizip.a  /usr/lib/x86_64-linux-gnu/libz.so  libgeometry.a  libbase.a  /usr/lib/x86_64-linux-gnu/libQt6Network.so.6.4.2  /usr/lib/x86_64-linux-gnu/libQt6Core.so.6.4.2  /usr/lib/x86_64-linux-gnu/libicui18n.so  /usr/lib/x86_64-linux-gnu/libfreetype.so  libvulkan_wrapper.a  libstb_image.a  /usr/lib/x86_64-linux-gnu/libharfbuzz.so.0.60830.0  /usr/lib/x86_64-linux-gnu/libGLX.so  /usr/lib/x86_64-linux-gnu/libOpenGL.so  libagg.a  /usr/lib/x86_64-linux-gnu/librt.a  -lm  -ldl && :
/usr/bin/ld.gold: error: cannot find -lqt_common
collect2: error: ld returned 1 exit status
```

I have all deps installed as per INSTALL.md including `qt6-base-dev`.

If then I try to build the `desktop` target (w/o clearing the output dir):
```
tools/unix/build_omim.sh -r desktop
...
[52/52] Linking CXX executable OMaps
FAILED: OMaps 
: && /usr/bin/ccache /usr/bin/c++ -O3 -DNDEBUG -fuse-ld=gold    -fdiagnostics-color=always qt/CMakeFiles/desktop.dir/Unity/unity_0_cxx.cxx.o qt/CMakeFiles/desktop.dir/qrc_resources.cpp.o -o OMaps  libgenerator.a  libqt_common.a  libmap.a  libgflags_nothreads.a  libtess2.a  librouting.a  libdescriptions.a  libkml.a  libdrape_frontend.a  libtransit.a  libtraffic.a  librouting_common.a  libshaders.a  libdrape.a  libsearch.a  libindexer.a  libeditor.a  libsearch.a  libindexer.a  libeditor.a  libstorage.a  libmwm_diff.a  libbsdiff.a  libge0.a  libopenlocationcode.a  libprotobuf.a  libopening_hours.a  libpugixml.a  libplatform.a  libcoding.a  libcppjansson.a  3party/jansson/jansson/lib/libjansson.a  libexpat.a  -lm  libsuccinct.a  /usr/lib/x86_64-linux-gnu/libicuuc.so  libminizip.a  /usr/lib/x86_64-linux-gnu/libz.so  libgeometry.a  libbase.a  /usr/lib/x86_64-linux-gnu/libQt6Network.so.6.4.2  /usr/lib/x86_64-linux-gnu/libQt6Positioning.so.6.4.2  /usr/lib/x86_64-linux-gnu/libicui18n.so  /usr/lib/x86_64-linux-gnu/libfreetype.so  libvulkan_wrapper.a  -ldl  libstb_image.a  /usr/lib/x86_64-linux-gnu/libharfbuzz.so.0.60830.0  libagg.a  /usr/lib/x86_64-linux-gnu/libQt6OpenGLWidgets.so.6.4.2  /usr/lib/x86_64-linux-gnu/libQt6Widgets.so.6.4.2  /usr/lib/x86_64-linux-gnu/libQt6OpenGL.so.6.4.2  /usr/lib/x86_64-linux-gnu/libQt6Gui.so.6.4.2  /usr/lib/x86_64-linux-gnu/libGLX.so  /usr/lib/x86_64-linux-gnu/libOpenGL.so  /usr/lib/x86_64-linux-gnu/libQt6Core.so.6.4.2 && :
libplatform.a(unity_0_cxx.cxx.o):unity_0_cxx.cxx:function QtLocationService::QtLocationService(location::LocationObserver&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&):(.text+0x18258): error: undefined reference to 'vtable for QtLocationService'
/usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
collect2: error: ld returned 1 exit status
```

If I build the desktop target first (with cleaning) then it builds and works fine (the binary get 9K bigger).
When I try to build `dev_sandbox` after the desktop:
```
[39/70] Building CXX object platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx.o
FAILED: platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx.o 
/usr/bin/ccache /usr/bin/ccache /usr/bin/c++ -DQT_CORE_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DRELEASE -DUSE_FILE32API -I/home/pastk/dev/OM/omim-build-release/platform/platform_autogen/include -I/home/pastk/dev/OM/organicmaps -I/home/pastk/dev/OM/omim-build-release/3party/gflags/include -I/home/pastk/dev/OM/organicmaps/3party/utfcpp/source -I/home/pastk/dev/OM/organicmaps/3party/expat/expat/lib -I/home/pastk/dev/OM/omim-build-release/3party/expat/expat -I/home/pastk/dev/OM/omim-build-release/3party/jansson/jansson/include -isystem /home/pastk/dev/OM/organicmaps/3party/boost -isystem /usr/include/x86_64-linux-gnu/qt6/QtCore -isystem /usr/include/x86_64-linux-gnu/qt6 -isystem /usr/lib/x86_64-linux-gnu/qt6/mkspecs/linux-g++ -isystem /usr/include/x86_64-linux-gnu/qt6/QtNetwork -O3 -DNDEBUG -std=c++20 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -ffast-math -Ofast -Wall -Wextra -Wpedantic -Wno-unused-parameter -fPIC -MD -MT platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx.o -MF platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx.o.d -o platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx.o -c /home/pastk/dev/OM/omim-build-release/platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx
In file included from /home/pastk/dev/OM/omim-build-release/platform/platform_autogen/EWIEGA46WW/moc_qt_location_service.cpp:10,
                 from /home/pastk/dev/OM/omim-build-release/platform/platform_autogen/mocs_compilation.cpp:3,
                 from /home/pastk/dev/OM/omim-build-release/platform/CMakeFiles/platform.dir/Unity/unity_0_cxx.cxx:4:
/home/pastk/dev/OM/omim-build-release/platform/platform_autogen/EWIEGA46WW/../../../../organicmaps/platform/qt_location_service.hpp:5:10: fatal error: QGeoPositionInfoSource: No such file or directory
    5 | #include <QGeoPositionInfoSource>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[40/70] Building CXX object search/CMakeFiles/search.dir/Unity/unity_1_cxx.cxx.o
ninja: build stopped: subcommand failed.
```

So there seems to be the same QtPositioning dependency bug as in 
- https://github.com/organicmaps/organicmaps/issues/9728
- potential fix https://github.com/organicmaps/organicmaps/pull/9981

renderexpert commented:
@pastk Thanks for checking! dev_sandbox should not be dependent on Qt, I'll handle it.

renderexpert commented:
> Also do we need to merge a related https://github.com/organicmaps/organicmaps/pull/9981 first?
@pastk Don't know yet, in theory dev_sandbox should not be dependent on Qt. So, if I can decouple them then they can be landed independently.

renderexpert commented:
@pastk Removed false dependancy to qt_common in dev_sandbox. Successfully repeated your test on my local Ubuntu machine

pastk commented:
Thanks! It builds now.
Though there is still the QtPositioning dependency issue when interleaving dev_sandbox and desktop builds. But its solved by https://github.com/organicmaps/organicmaps/pull/9981

pastk commented:
There is a blocker though:
it makes everything look bleak both in the Qt desktop app and in the sandbox app. Like there is a semi-transparent overlay.

![2025-02-27_22-45_bleak](https://github.com/user-attachments/assets/0ceeb0d7-14aa-4ae7-962c-c295a24da07f)

Actually its supposed to be like
![2025-02-27_22-46_normal](https://github.com/user-attachments/assets/e42263f3-fcfd-47c7-803e-50a84fbf2d85)

In the sandbox app the issue goes away after switching the renderer (to opengl; switching back to vulkan is fine also).
There is no such option in the Qt app.

Could be caused by a newly added shader?

pastk commented:
Smaller issues that could be fixed later:
- https://github.com/organicmaps/organicmaps/issues/10361
- https://github.com/organicmaps/organicmaps/issues/10362

renderexpert commented:
@pastk Thanks!

> Could be caused by a newly added shader?

I doubt it, Qt desktop app and dev_sandbox are independent app. ImGUI exists only in dev_sandbox, so can't affect on Qt desktop app. If it happens for Qt desktop app too, did you check on master without this PR?

pastk commented:
The master branch is fine. The bug is present on this PR branch only.

renderexpert commented:
@pastk Ok, I reproduced problem on Ubuntu in dev_sandbox. It can't be explained different surface formats in Vulkan and OpenGL (one may be SRBG, another is RGB). However, I can't repro the same behaviour in the Qt app (PR vs master). Are you sure that you rebuilded everything when you checked Qt app?

PR:
![Screenshot from 2025-02-28 23-24-42](https://github.com/user-attachments/assets/18089635-30cf-4a87-bcc5-82de6b644d4c)

Master:
![Screenshot from 2025-02-28 23-38-51](https://github.com/user-attachments/assets/7d8420ce-c0a0-415a-bdc7-8646f890fa26)

renderexpert commented:
@pastk I added commit that may fix your issue in dev_sandbox (again I can't repro the same in the Qt App).
If it doesn't help you, I have to ask your help in putting breakpoint here
```
organicmaps/drape/vulkan/vulkan_context_factory.cpp:366
```
It should be something like that. I need to know what format was chosen. Thanks!
![Screenshot from 2025-03-01 00-09-35](https://github.com/user-attachments/assets/c197e581-eaf4-4324-8b60-7c046cdcbc83)

pastk commented:
I'm sorry, its my bad, the gamma correction commit is not needed.
I had the subway layer ON and didn't notice it :)

I thought that its a dev_sandbox bug because switching the rendering API disables the layer (and switching back to Vulkan doesn't re-enable it).

So the actual bug is that dev sandbox initially inherits the layer status from the Qt app, but resets it upon API change, which is unexpected and confusing.

renderexpert commented:
@rtsisyk @vng It looks like the issue resolved. Can we merge this PR?

vng commented:
@pastk Approve? Since you had some Linux questions here ..
